pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -475,14 +482,35 @@ func (nc NodeController) tryUpdateNodeStatus(node *api.Node) (error, time.Durati
 			LastTransitionTime: node.CreationTimestamp,
 		}
 		gracePeriod = nodeStartupGracePeriod
+		nc.nodeStatusMap[node.Name] = NodeStatusData{
+			status:    node.Status,
+			timestamp: node.CreationTimestamp,
+		}
 	} else {
 		// If ready condition is not nil, make a copy of it, since we may modify it in place later.
 		lastReadyCondition = *readyCondition
 		gracePeriod = nodeMonitorGracePeriod
 	}
 
-	// Check last time when NodeReady was updated.
-	if nc.now().After(lastReadyCondition.LastProbeTime.Add(gracePeriod)) {
+	oldNodeStatus, found := nc.nodeStatusMap[node.Name]
+	if !found {
+		glog.Errorf("NodeController state corrupted: timestamp map misses entry for %s. Assuming now.", node.Name)
```

Comment:
I don't think this is an error condition. Because NodeStatus is stored in etcd but the map is in-memory, whenever you restart the master, I think you will hit this condition the first time you loop over each node. So I would change it to an Infof.